events/keywords.py

```diff
@@ -1,112 +1,78 @@
+            # storing the result in a dictionary of the following structure {similarity: label}
+            # in the edge case when similarity is the same for two different languages the label will be
+            # overwritten, which is not big deal as we don't have a way to select between the two anyway.
+            if label:
+                contestants[label.similarity] = label
 
-        if not matches:
-            print('no match', text)
+        # selecting the match with the highest similarity, if there is anything to select
+        if contestants.keys():
+            return contestants[max(contestants.keys())]
+        else:
             return None
 
-        keyword_ids = set()
-        if match_type not in ['exact', 'subword']:
-            cmatch = get_close_matches(
-                text, [m.lower() for m in matches], n=1)
-            if len(cmatch) == 1:
-                keyword_ids = self.name_to_keyword_ids.get(cmatch[0])
+    def label_match(self, text, language=None):
+        # Looking for an exact match regardless of the language
+        label = KeywordLabel.objects.filter(name__iexact=text)
+        if label:
+            return label
 
-        else:
-            for m in matches:
-                keyword_ids.update(self.name_to_keyword_ids[m])
+        # If no exact matches found, let's use Postgres full-text search
+        # to find a label matched by lexeme and rank the results with
+        # TrigramSimilarity as ft SearchRank is not suitable for ranking matched
+        # individual words. If no language is passed we cycle through all
+        # the options as specified in FULLTEXT_SEARCH_LANGUAGES and select
+        # the best match according to similarity.
+        label = self.full_text_matching(text, language)
+        if label:
+            return [label]
 
-        if len(keyword_ids) < 1:
-            print('no matches for', text)
+        # if no matches found let's check if we could split the string
+        texts = text.split()
+        if len(texts) > 1:
+            all_the_labels = []
+            for word in texts:
+                label = self.label_match(word, language)
+                if label:
+                    all_the_labels.extend(label)
+            return all_the_labels if all_the_labels else None
+        else:
             return None
 
-        objects = Keyword.objects.filter(id__in=keyword_ids, deprecated=False)
-        if len(keyword_ids) > 1:
-            try:
-                aggregate_keyword = objects.get(aggregate=True)
-                aggregate_name = re.sub(ENDING_PARENTHESIS_PATTERN, '', aggregate_keyword.name_fi)
-                result = [aggregate_keyword]
-                for o in objects.exclude(name_fi__istartswith=aggregate_name):
-                    result.append(o)
-                return result
-            except Keyword.DoesNotExist:
-                pass
-            return objects
-        return objects
+    def match(self, text, language=None):
```

Comment:
If any, this needs unit tests, since label_match is a self created unknown and untested method.